File: service/src/main/java/com/ricardococati/service/impl/RecomendacaoSemanalCalculaServiceImpl.java
package com.ricardococati.service.impl;

import static com.ricardococati.model.enums.Decisao.COMPRA;
import static com.ricardococati.model.enums.Decisao.NEUTRO;
import static com.ricardococati.model.enums.Decisao.VENDE;
import static java.util.Objects.nonNull;

import com.ricardococati.model.entities.RecomendacaoSemanal;
import com.ricardococati.repository.dao.HistogramaSemanalInserirDAO;
import com.ricardococati.repository.dao.MacdSemanalBuscarDAO;
import com.ricardococati.repository.dao.MediaMovelExponencialSemanalBuscarDAO;
import com.ricardococati.repository.dao.RecomendacaoSemanalBuscarDAO;
import com.ricardococati.repository.dao.RecomendacaoSemanalExcluirDAO;
import com.ricardococati.repository.dao.RecomendacaoSemanalInserirDAO;
import com.ricardococati.repository.dao.SinalMacdSemanalBuscarDAO;
import com.ricardococati.service.CalculaService;
import com.ricardococati.service.CandlestickSemanalBuscarService;
import com.ricardococati.service.RecomendacaoSemanalCalculaService;
import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.stream.IntStream;
import lombok.Data;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;

@Slf4j
@Data
@Service
@RequiredArgsConstructor
public class RecomendacaoSemanalCalculaServiceImpl
    implements RecomendacaoSemanalCalculaService {

  private final CandlestickSemanalBuscarService calculaCandlestickService;
  private final MacdSemanalBuscarDAO macdDAO;
  private final SinalMacdSemanalBuscarDAO sinalMacdDAO;
  private final HistogramaSemanalInserirDAO histogramaDAO;
  private final RecomendacaoSemanalBuscarDAO buscarRecomendacao;
  private final RecomendacaoSemanalInserirDAO inserirRecomendacao;
  private final RecomendacaoSemanalExcluirDAO excluirRecomendacao;
  private final MediaMovelExponencialSemanalBuscarDAO mediaMovelExponencialDAO;
  private final CalculaService calculaService;

  @Override
  public List<RecomendacaoSemanal> executeByCodNeg(
      final List<String> listCodneg, final LocalDate dtLimitePregao) {
    excluirRecomendacao.excluirAllRecomendacao();
    List<RecomendacaoSemanal> diarioList = new ArrayList<>();
    log.info("Código de negociação: " + listCodneg);
    listCodneg
        .parallelStream()
        .filter(Objects::nonNull)
        .forEachOrdered(codneg -> {
          diarioList.addAll(calculaRecomendacao(codneg, dtLimitePregao));
          incluirRecomendacao(diarioList);
        });
    return diarioList;
  }

  private void incluirRecomendacao(final List<RecomendacaoSemanal> semanalList) {
    semanalList
        .stream()
        .filter(Objects::nonNull)
        .filter(mmsSemanal -> nonNull(mmsSemanal.getDtpregini()))
        .filter(mmsSemanal -> nonNull(mmsSemanal.getDtpregfim()))
        .filter(mmsSemanal -> nonNull(mmsSemanal.getRecomendacao()))
        .filter(mmsSemanal -> nonNull(mmsSemanal.getRecomendacao().getCodneg()))
        .forEach(inserirRecomendacao::incluirRecomendacao);
  }

  private List<RecomendacaoSemanal> calculaRecomendacao(
      final String codneg, final LocalDate dtLimitePregao) {
    List<RecomendacaoSemanal> recomendacaoList = buildListRecomendacao(
        codneg,
        dtLimitePregao
    );
    IntStream.range(0, recomendacaoList.size())
        .filter(indice -> indice > 0 && recomendacaoList.size() >= 2)
        .forEach(indice -> {
          if (recomendacaoList.get(indice - 1).getRecomendacao().getPrecoHistograma()
              .compareTo(recomendacaoList.get(indice).getRecomendacao().getPrecoHistograma()) < 0) {
            recomendacaoList.get(indice).getRecomendacao().setDecisao(COMPRA.getTexto());
          } else if (recomendacaoList.get(indice - 1).getRecomendacao().getPrecoHistograma()
              .compareTo(recomendacaoList.get(indice).getRecomendacao().getPrecoHistograma()) > 0) {
            recomendacaoList.get(indice).getRecomendacao().setDecisao(VENDE.getTexto());
          } else {
            recomendacaoList.get(indice).getRecomendacao().setDecisao(NEUTRO.getTexto());
          }
        });
    return recomendacaoList;
  }

  private List<RecomendacaoSemanal> buildListRecomendacao(
      final String codneg,
      final LocalDate dtLimitePregao
  ) {
    return buscarRecomendacao.getListRecomendacaoByDtPregECodNeg(
        dtLimitePregao,
        codneg
    );
  }

}

File: service/src/main/java/com/ricardococati/service/impl/RecomendacaoSemanalCalculaServiceImpl.kt
package com.ricardococati.service.impl

import com.ricardococati.model.entities.RecomendacaoSemanal
import com.ricardococati.model.enums.Decisao
import com.ricardococati.repository.dao.HistogramaSemanalInserirDAO
import com.ricardococati.repository.dao.MacdSemanalBuscarDAO
import com.ricardococati.repository.dao.MediaMovelExponencialSemanalBuscarDAO
import com.ricardococati.repository.dao.RecomendacaoSemanalBuscarDAO
import com.ricardococati.repository.dao.RecomendacaoSemanalExcluirDAO
import com.ricardococati.repository.dao.RecomendacaoSemanalInserirDAO
import com.ricardococati.repository.dao.SinalMacdSemanalBuscarDAO
import com.ricardococati.service.CalculaService
import com.ricardococati.service.CandlestickSemanalBuscarService
import com.ricardococati.service.RecomendacaoSemanalCalculaService
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import java.time.LocalDate
import java.util.ArrayList
import java.util.Objects
import java.util.stream.IntStream

@Service
class RecomendacaoSemanalCalculaServiceImpl(
        val calculaCandlestickService: CandlestickSemanalBuscarService,
        val macdDAO: MacdSemanalBuscarDAO,
        val sinalMacdDAO: SinalMacdSemanalBuscarDAO,
        val histogramaDAO: HistogramaSemanalInserirDAO,
        val buscarRecomendacao: RecomendacaoSemanalBuscarDAO,
        val inserirRecomendacao: RecomendacaoSemanalInserirDAO,
        val excluirRecomendacao: RecomendacaoSemanalExcluirDAO,
        val mediaMovelExponencialDAO: MediaMovelExponencialSemanalBuscarDAO,
        val calculaService: CalculaService
) : RecomendacaoSemanalCalculaService {

    override fun executeByCodNeg(
            listCodneg: List<String>, dtLimitePregao: LocalDate): List<RecomendacaoSemanal> {
        excluirRecomendacao.excluirAllRecomendacao()
        val diarioList: MutableList<RecomendacaoSemanal> = ArrayList()
        log.info("Código de negociação: $listCodneg")
        listCodneg
                .parallelStream()
                .filter { obj: String? -> Objects.nonNull(obj) }
                .forEachOrdered { codneg: String ->
                    diarioList.addAll(calculaRecomendacao(codneg, dtLimitePregao))
                    incluirRecomendacao(diarioList)
                }
        return diarioList
    }

    private fun incluirRecomendacao(semanalList: List<RecomendacaoSemanal>) {
        semanalList
                .stream()
                .filter { obj: RecomendacaoSemanal? -> Objects.nonNull(obj) }
                .filter { mmsSemanal: RecomendacaoSemanal -> Objects.nonNull(mmsSemanal.dtpregini) }
                .filter { mmsSemanal: RecomendacaoSemanal -> Objects.nonNull(mmsSemanal.dtpregfim) }
                .filter { mmsSemanal: RecomendacaoSemanal -> Objects.nonNull(mmsSemanal.recomendacao) }
                .filter { mmsSemanal: RecomendacaoSemanal -> Objects.nonNull(mmsSemanal.recomendacao!!.codneg) }
                .forEach { recomendacaoSemanal: RecomendacaoSemanal? -> inserirRecomendacao.incluirRecomendacao(recomendacaoSemanal) }
    }

    private fun calculaRecomendacao(
            codneg: String, dtLimitePregao: LocalDate): List<RecomendacaoSemanal> {
        val recomendacaoList = buildListRecomendacao(
                codneg,
                dtLimitePregao
        )
        IntStream.range(0, recomendacaoList.size)
                .filter { indice: Int -> indice > 0 && recomendacaoList.size >= 2 }
                .forEach { indice: Int ->
                    if (recomendacaoList[indice - 1].recomendacao!!.precoHistograma
                                    ?.compareTo(recomendacaoList[indice].recomendacao!!.precoHistograma)!! < 0) {
                        recomendacaoList[indice].recomendacao!!.decisao = Decisao.TENDENCIA_ALTA.texto
                    } else if (recomendacaoList[indice - 1].recomendacao!!.precoHistograma
                                    ?.compareTo(recomendacaoList[indice].recomendacao!!.precoHistograma)!! > 0) {
                        recomendacaoList[indice].recomendacao!!.decisao = Decisao.TENDENCIA_BAIXA.texto
                    } else {
                        recomendacaoList[indice].recomendacao!!.decisao = Decisao.CONSOLIDANDO.texto
                    }
                }
        return recomendacaoList
    }

    private fun buildListRecomendacao(
            codneg: String,
            dtLimitePregao: LocalDate
    ): List<RecomendacaoSemanal> {
        return buscarRecomendacao.getListRecomendacaoByDtPregECodNeg(
                dtLimitePregao,
                codneg
        )
    }

    override fun toString(): String {
        return ("RecomendacaoSemanalCalculaServiceImpl(calculaCandlestickService=" + calculaCandlestickService + ", macdDAO=" + macdDAO + ", sinalMacdDAO="
                + sinalMacdDAO + ", histogramaDAO=" + histogramaDAO
                + ", buscarRecomendacao=" + buscarRecomendacao + ", inserirRecomendacao=" + inserirRecomendacao + ", excluirRecomendacao=" + excluirRecomendacao
                + ", mediaMovelExponencialDAO=" + mediaMovelExponencialDAO + ", calculaService="
                + calculaService + ")")
    }

    companion object {
        private val log = LoggerFactory
                .getLogger(RecomendacaoSemanalCalculaServiceImpl::class.java)
    }

}